segbot_sensors/src/segbot_sensors/battery_monitor.cpp

```diff
@@ -0,0 +1,127 @@
+#include "ros/time.h"
+#include <vector>
+#include "std_msgs/Header.h"
+#include "std_msgs/Float32.h"
+#include "diagnostic_msgs/DiagnosticStatus.h"
+#include "diagnostic_msgs/DiagnosticArray.h"
+#include "diagnostic_msgs/KeyValue.h"
+#include "smart_battery_msgs/SmartBatteryStatus.h" //file found
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+float voltage;
+bool sentMail = false;
+void voltageCallback(const smart_battery_msgs::SmartBatteryStatus::ConstPtr& msg)
+{
+  voltage = (msg->voltage);
+ // ROS_INFO("I heard: [%f]", msg->voltage);
+  //todo: add voltage profiler logic - write profile rates in /config
+}
+
+int sendMail()
+{
+    char msg[] = "Robot battery low. Please seek robot and bring to lab for charging.\nNOTE:This is an automated message generated by the diagnostic aggregator node. Do NOT reply to this message.";
+    int retval = -1;
+    FILE *mailpipe = popen("/usr/lib/sendmail -t", "w");
+    if (mailpipe != NULL) {
+        fprintf(mailpipe, "To: maxsvetlik@gmail.com\n");
+        fprintf(mailpipe, "From: BWIrobotAlert@bwi.utexas.edu\n");
+        fprintf(mailpipe, "Subject: Segbot Low Battery Alert\n");
+        fwrite(msg, 1, strlen(msg), mailpipe);
+        fwrite(".\n", 1, 2, mailpipe);
+        pclose(mailpipe);
+        retval = 0;
+     }
+     else {
+         perror("Failed to invoke sendmail");
+     }
+     return retval;
+}
+
+
+int main(int argc, char **argv)
+{
+  ros::init(argc, argv, "battery_estimator");
+  ros::NodeHandle n;
+  ros::Publisher battery_life_pub = n.advertise<diagnostic_msgs::DiagnosticArray>("diagnostics", 10);
+  ros::Subscriber voltage_sub = n.subscribe("smart_battery_msgs/SmartBatteryStatus", 10, voltageCallback);
+  ros::Rate loop_rate(1); //1hz
+  diagnostic_msgs::DiagnosticStatus voltages;
+  diagnostic_msgs::DiagnosticStatus status;
+  diagnostic_msgs::DiagnosticArray diagAr;
+  diagnostic_msgs::KeyValue status_val;
+  diagnostic_msgs::KeyValue voltages_val;
+  voltages.hardware_id = "005";
+  voltages.name = "voltage";
+  voltages_val.value = voltage;
+  voltages_val.key = "Current voltage: ";
+  status.hardware_id = "005";
+  status.name = "battery_estimator"; //must match what the aggregator config file expects
+  status_val.key = "battery_value";
+  status_val.value = "183.0";
+  std::ostringstream ss;
+
+  while (ros::ok())
+  {
+    //conver from float to string - ugly
+    ss.str("");
+    ss.clear();
+    ss << voltage;
+    voltages_val.value  = ss.str();
+    diagAr.header.stamp = ros::Time::now();
+    diagAr.header.frame_id = 1;
+    //float estimated_life = 183.0;
+    
+    if(voltage > 11.0){
+		status.message = "Battery in good health";
+		status.level = 0; // 0 = OK
+		voltages.level = 0;
+		voltages.message = "Voltage level OK";
+    }
+    else if(voltage > 10 && voltage < 11){
+		status.message = "Battery low, return to lab.";
+		status.level = 1; // WARN
+		voltages.level = 1;
+		voltages.message = "Voltage dropping";
+		if(!sentMail){
```

Comment:
This might be a good time for you to lookup boost::thread. Forking a new process here is overkill. You can simply call the sendMail function in a new thread.